Persephone/Controllers/QueueViewController.swift

```diff
@@ -0,0 +1,183 @@
+//
+//  QueueController.swift
+//  Persephone
+//
+//  Created by Daniel Barber on 2019/2/01.
+//  Copyright © 2019 Dan Barber. All rights reserved.
+//
+
+import Cocoa
+import mpdclient
+
+class QueueViewController: NSViewController, NSOutlineViewDataSource, NSOutlineViewDelegate {
+  var queue: [MPDClient.Song] = []
+  var queuePos: Int32 = -1
+
+  var queueIcon: NSImage? = nil
+
+  let systemFontRegular = NSFont.systemFont(ofSize: 13, weight: .regular)
+  let systemFontBold = NSFont.systemFont(ofSize: 13, weight: .bold)
+
+  struct SongItem {
+    var song: MPDClient.Song
+    var queuePos: Int
+  }
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    queueView.columnAutoresizingStyle = .uniformColumnAutoresizingStyle
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(stateChanged(_:)),
+      name: MPDClient.stateChanged,
+      object: AppDelegate.mpdClient
+    )
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(queueChanged(_:)),
+      name: MPDClient.queueChanged,
+      object: AppDelegate.mpdClient
+    )
+
+    NotificationCenter.default.addObserver(
```

Comment:
Do these observers need to be removed at some point in the app's lifecycle?

Reply:
Oh quite possibly. Does Swift not handle that automatically if the calling class is de-instantiated?

Reply:
To be honest, I don't know with the most recent versions of Swift. I've just been conditioned to always remove them by crashers in ObjC that happened because a listener tried to fire on an object that was no longer there.

Reply:
Because these are using the [selector-based observer API](https://developer.apple.com/documentation/foundation/notificationcenter/1415360-addobserver), they are automatically removed when the observer is deallocated:

> If your app targets iOS 9.0 and later or macOS 10.11 and later, you don't need to unregister an observer in its `dealloc` method. Otherwise, you should call `removeObserver(_:name:object:)` before observer or any object passed to this method is deallocated.

Reply:
Looks like this isn't necessary anymore?

https://developer.apple.com/documentation/foundation/notificationcenter/1413994-removeobserver

Reply:
Ha! @sharplet beat me to it.